### PR TITLE
[testing]: Add "orca_per_rpc" test case

### PR DESCRIPTION
**Pull request by eugeneo**

### CODE REVIEW

File: test/cpp/interop/interop_client.cc
Comment (markdroth) [inline test/cpp/interop/interop_client.cc:162]: `std::move()`

File: test/cpp/interop/interop_client.h
Comment (markdroth) [inline test/cpp/interop/interop_client.h:45]: Instead of exposing this type and passing it into the ctor, why not just register the LB policy inside the ctor and have the callback directly write to the data member of `InteropClient`, so that it doesn't need to be a `shared_ptr<>`?  Even if you need this behavior to be configurable for the Android client, you can pass in a bool indicating whether to configure the LB policy.

File: test/cpp/interop/interop_server.cc
Comment (markdroth) [inline test/cpp/interop/interop_server.cc:155]: Are you sure this works correctly?  This is storing `absl::string_view` keys in the `CallMetricRecorder` that point to storage in the request proto, but I don't think that request proto is guaranteed to live until the ORCA metadata is serialized to be sent back to the client, so it seems like this could leave pointers dangling into unsafe memory.

In our e2e tests, we deal with this by arena-allocating a copy of these strings:

https://github.com/grpc/grpc/blob/c5213ceff9edc723cbf56a5f917083ff4a025bf0/test/cpp/end2end/client_lb_end2end_test.cc#L141

File: test/cpp/interop/interop_server.cc
Comment (markdroth) [inline test/cpp/interop/interop_server.cc:389]: In this particular case, we don't need a `ServerMetricRecorder`, so I don't think it's necessary to create this.  You can just pass null to `EnableCallMetricRecording()`.

**Review by markdroth — COMMENTED** [review]
Thanks for doing this.  This looks really good overall!

My main comment is the one about exposing the type of the map to store the backend metrics in.  I think we can just do that internally inside of `InteropClient`, so that it doesn't need to be exposed to callers.

Please let me know if you have any questions.  Thanks!

File: src/proto/grpc/testing/BUILD
Comment (eugeneo) [reply src/proto/grpc/testing/BUILD:210]: Done

File: test/cpp/interop/client.cc
Comment (eugeneo) [reply test/cpp/interop/client.cc:128]: I like optionals - compiler makes sure I don't forget to check if it is set...

I switched to string.

File: test/core/util/test_lb_policies.h
Comment (eugeneo) [reply test/core/util/test_lb_policies.h:70]: This policy is referred by name [here](https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md#orca_per_rpc) - I am assuming this config will be provided from outside. I did entertain a though of changing the default name for the policy, lmk if you'd rather have that.

File: test/cpp/interop/client_helper.cc
Comment (eugeneo) [reply test/cpp/interop/client_helper.cc:91]: It would require `ABSL_FLAG` in each client, that would make this PR much bigger. Another option would be to have that declaration in `client_helper.cc` which seems to be discouraged.

File: src/android/test/interop/app/src/main/cpp/grpc-interop.cc
Comment (eugeneo) [reply src/android/test/interop/app/src/main/cpp/grpc-interop.cc:44]: I am not familiar with Android at this point. This line only initializes the custom LB policy.

File: test/cpp/interop/interop_client.h
Comment (eugeneo) [reply test/cpp/interop/interop_client.h:45]: `serviceStub_` creation opens a channel, so LB policy needs to be created before that one is created. I considered multiple approaches but then decided that passing it here will offer more flexibility as I am not familiar enough with all the tests that use this code.

File: src/android/test/interop/app/src/main/cpp/grpc-interop.cc
Comment (eugeneo) [reply src/android/test/interop/app/src/main/cpp/grpc-interop.cc]: Reverted

File: test/core/util/BUILD
Comment (eugeneo) [reply test/core/util/BUILD:292]: There are more executables than just interop_client. I am not sure what are the implications of marking them all as `testonly`.

File: test/cpp/interop/interop_client.h
Comment (eugeneo) [reply test/cpp/interop/interop_client.h]: Please take a look at a new implementation

File: test/cpp/interop/interop_server.cc
Comment (eugeneo) [reply test/cpp/interop/interop_server.cc]: Copied. Was not familiar with this API.

File: test/cpp/interop/interop_server.cc
Comment (eugeneo) [reply test/cpp/interop/interop_server.cc]: Done.

**Review by eugeneo — COMMENTED** [review]
Thank you for the review. Please take another look.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [reply test/cpp/interop/interop_client.h]: Thanks, I like this a lot better, since the ugliness is now hidden inside the client code.

In the long run, I think we should probably change this LB policy to get the address of the callback from a channel arg instead of it being globally set at LB policy registration time.  But that doesn't need to block this PR.

**Comment by eugeneo** [comment]
This was an extensive change. I did a full reset to revert the files that no longer need to be changed. Please take another look.

I don't think failed checks should block the review at this point.

File: test/cpp/interop/interop_client.h
Comment (eugeneo) [reply test/cpp/interop/interop_client.h]: Done

File: test/cpp/interop/client_helper.cc
Comment (eugeneo) [reply test/cpp/interop/client_helper.cc]: No longer needed. Clients need to configure all args.

File: test/core/util/test_lb_policies.h
Comment (eugeneo) [reply test/core/util/test_lb_policies.h]: Now this is a distinct LB policy

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.cc:78]: We need to remove this pollset_set linkage in the dtor.

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.cc:73]: Suggest moving this to the top of the function, before you start to build delegate_args, since it's really not related to constructing the delegate policy.  Note that it's fine to call `channel_args()` twice, since that method returns a const reference.

Also, please add an assertion that `load_report_tracker_` is non-null.

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.cc:222]: I don't think we need to check whether it's null here, because `BackendMetricDataToOrcaLoadReport()` will return nullopt in that case.  I think we can just unconditionally add the result of `BackendMetricDataToOrcaLoadReport()` to the queue.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [inline test/cpp/interop/interop_client.h:125]: Shouldn't this still use `ChannelCreationFunc` rather than spelling out the type?

File: test/cpp/interop/interop_client.h
Comment (markdroth) [inline test/cpp/interop/interop_client.h:135]: Same here.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [inline test/cpp/interop/interop_client.h:170]: I don't think we need the "Needs to be initialized first" comment anymore.

File: test/cpp/interop/client.cc
Comment (markdroth) [inline test/cpp/interop/client.cc:245]: `std::move()`

File: test/cpp/interop/interop_client.h
Comment (markdroth) [inline test/cpp/interop/interop_client.h:43]: Instead of using a function that takes another function as a parameter, I think it would be simpler to just pass in the channel args from the caller, like this:

```
typedef std::function<std::shared_ptr<Channel>(ChannelArguments)>
    ChannelCreationFuncWithCustomArgs;
```

Then the new `InteropClient` ctor can look like this:

```
InteropClient::InteropClient(
    ChannelCreationFuncWithCustomArgs channel_creation_func,
    bool new_stub_every_test_case, bool do_not_abort_on_transient_failures)
    : serviceStub_(
          [&]() {
            InitializeCustomLbPolicyIfNeeded();
            ChannelArguments args;
            load_report_tracker_->SetupOnChannel(&args);
            return channel_creation_func(std::move(args));
          },
          new_stub_every_test_case),
      do_not_abort_on_transient_failures_(do_not_abort_on_transient_failures) {}
```

And the lambdas in client.cc can just add their own args to the passed-in args.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [inline test/cpp/interop/interop_client.h:57]: Instead of adding a second ctor, why not just change the existing callers to use the new parameter?

File: test/cpp/interop/client.cc
Comment (markdroth) [inline test/cpp/interop/client.cc:217]: `std::move()`

**Review by markdroth — COMMENTED** [review]
The new LB policy looks great!

Just one substantive comment here, which is the one about simplifying the way that channel args are passed through the interop code.  The current approach seems more complicated than it needs to be, and it took me a while to understand.

Please let me know if you have any questions.  Thanks!

File: test/cpp/interop/backend_metrics_lb_policy.h
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.h]: Done

File: test/cpp/interop/backend_metrics_lb_policy.h
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.h]: Done

Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.h]: Done

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.cc]: Done (it lives in the header now)

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.cc]: Done

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.cc:75]: Done, as I understand it... Went to ShutdownLocked, as that's where the delegate is reset.

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.cc]: Done

**Review by eugeneo — COMMENTED** [review]
Done, please take another look.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [reply test/cpp/interop/interop_client.h:125]: Looks like you made this change below but not here in the argument to the `ServiceStub` ctor.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [reply test/cpp/interop/interop_client.h]: I didn't mean to remove the whole comment, just the second sentence.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [reply test/cpp/interop/interop_client.h:43]: I don't understand what you mean about not wanting to bury the service config flag.  The approach I proposed here would not require changing where the flag is handled; that would still happen in the lambdas that are created in client.cc.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [reply test/cpp/interop/interop_client.h:57]: I only see 3 other callers, and they look trivial to change.  Avoiding this doesn't seem to save much time.  And unless you plan to write a subsequent PR to clean this up, this is just leaving unnecessary technical debt.

The down-side of this approach is that any caller that uses the old interface will not get the `LoadReportTracker` channel arg.

Comment (eugeneo) [reply test/cpp/interop/interop_client.h]: Every caller also needs to be retrofitted with the JSON command line arg and such. I understand this is not necessary for current pull request and figuring out all the build dependencies and such delays this pull request.

I removed the second ctor but would add JSON config flag in a separate PR if needed. I feel like some refactoring effort is needed to reduce code duplication.

File: test/cpp/interop/backend_metrics_lb_policy.h
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.h]: Done

File: test/cpp/interop/interop_client.h
Comment (eugeneo) [reply test/cpp/interop/interop_client.h]: Done

File: test/cpp/interop/interop_client.h
Comment (eugeneo) [reply test/cpp/interop/interop_client.h]: Done

File: test/cpp/interop/backend_metrics_lb_policy.h
Comment (eugeneo) [reply test/cpp/interop/backend_metrics_lb_policy.h]: Done

**Review by eugeneo — COMMENTED** [review]
Thank you for the review. I addressed the comments. PTAL.

File: test/cpp/interop/backend_metrics_lb_policy.h
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.h:32]: The style guide says not to use this kind of "using" statement in a .h file:

https://google.github.io/styleguide/cppguide.html#Namespaces

File: test/cpp/interop/backend_metrics_lb_policy.h
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.h:40]: I don't think we need a `StatusOr<>` here, just an `optional<>`.  I realize that makes two levels of optionals, which is a little awkward, but I think you can make it clearer with a typedef:

```
// A load report, or nullopt if the call had no load report.
using LoadReportEntry = absl::optional<xds::data::orca::v3::OrcaLoadReport>;

// Returns the next load report, or nullopt if the queue is empty.
absl::optional<LoadReportEntry> GetNextLoadReport();
```

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.cc:232]: `std::move()`

File: test/cpp/interop/backend_metrics_lb_policy.cc
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.cc:30]: Suggest moving this down next to the other constant definition, on line 38.

File: test/cpp/interop/client.cc
Comment (markdroth) [inline test/cpp/interop/client.cc:231]: This should still use `std::move()`.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [inline test/cpp/interop/interop_client.h:114]: This should be public within `ServiceStub`.

File: test/cpp/interop/interop_client.h
Comment (markdroth) [reply test/cpp/interop/interop_client.h]: I don't think the JSON command line flag is needed in the other binaries, so I don't think there's any need for a follow-on PR here.  This looks great!

**Review by markdroth — APPROVED** [review]
This looks really good!

Only minor comments remaining; feel free to merge after addressing.

Please let me know if you have any questions.  Thanks!

**Review by eugeneo — COMMENTED** [review]
Thank you for the review. I will work on cherry-picking.

File: test/cpp/interop/backend_metrics_lb_policy.h
Comment (markdroth) [inline test/cpp/interop/backend_metrics_lb_policy.h:44]: This can be `std::deque<LoadReportEntry>`.

File: test/cpp/interop/client.cc
Comment (eugeneo) [reply test/cpp/interop/client.cc]: clang-tidy does not allow it...
```
test/cpp/interop/client.cc:231:41: error: passing result of std::move() as a const reference argument; no move will actually happen [performance-move-const-arg,-warnings-as-errors]
                                        std::move(arguments));
                                        ^~~~~~~~~~         ~
Suppressed 9286 warnings (9210 in non-user code, 76 NOLINT).
```

**Review by markdroth — APPROVED** [review]

File: test/cpp/interop/client.cc
Comment (markdroth) [reply test/cpp/interop/client.cc:231]: Ah, I didn't realize that `CreateChannelForTestCase()` takes the argument as a const reference.  It might make sense to change it to do so, but the way you have it now is fine too.

**Review by markdroth — APPROVED** [review]

**Comment by stanley-cheung** [comment]
@eugeneo Hi, this PR might have broken the observability interop test. Those tests are ready to be run as PR tests yet so there wasn't a signal there. I noticed that the observability tests start failing after this PR is merged. Here's the logs that fails:

https://fusion2.corp.google.com/ci/kokoro/prod:grpc-gcp%2Ftools%2Fobservability%2Fmaster%2Fcontinuous_cpp/activity/f1566fe2-1017-4514-b828-c1c28be2ebf2/invocations/f1566fe2-1017-4514-b828-c1c28be2ebf2/targets/integration%2Fcpp:cpp:test_monitoring_basic/log

The observability tests also uses the interop server/client. I also see you make some changes to `observability_client.cc` as well. 

Do you mind taking a look? Hopefully it's something obvious. If it's not, I can help set up some way for you to reproduce this. It's basically just docker pull an image from `gcr.io`, set some environment variables and run a `docker run` command. 

Thanks.

**Comment by stanley-cheung** [comment]
> It is fairly obvious. I will try to fix it.

Thanks! Let me know if you have a PR, I can run it through.

**Comment by eugeneo** [comment]
@stanley-cheung #32630 is in review now.

**Comment by stanley-cheung** [comment]
Thanks, I just kicked off an observability interop test run. Will report back. Thanks!

**Comment by eugeneo** [comment]
That PR did not build, sorry. Please wait for checks to pass. Thanks!

**Comment by stanley-cheung** [comment]
Ah, cool, will wait

**Comment by eugeneo** [comment]
@stanley-cheung that PR was checked in.

**Comment by stanley-cheung** [comment]
@eugeneo That fixed the issue! Thanks a lot!

**Pull request merged**
